go/vt/sqlparser/ast_format.go

```diff
@@ -1359,6 +1359,39 @@ func (node *Literal) Format(buf *TrackedBuffer) {
 
 // Format formats the node.
 func (node *Argument) Format(buf *TrackedBuffer) {
+	// We need to make sure that any value used still returns
+	// the right type when interpolated. For example, if we have a
+	// decimal type with 0 scale, we don't want it to be interpreted
+	// as an integer after interpolation as that would the default
+	// literal interpretation in MySQL.
+	switch {
+	case node.Type == sqltypes.Unknown:
+		// Ensure we handle unknown first as we don't want to treat
+		// the type as a bitmask for the further tests.
+		// do nothing, the default literal will be correct.
+	case sqltypes.IsDecimal(node.Type):
+		buf.astPrintf(node, "CAST(:%#s AS DECIMAL(%d, %d))", node.Name, node.Size, node.Scale)
+		return
+	case sqltypes.IsUnsigned(node.Type):
+		buf.astPrintf(node, "CAST(:%#s AS UNSIGNED)", node.Name)
+		return
+	case node.Type == sqltypes.Float64:
+		buf.astPrintf(node, "CAST(:%#s AS DOUBLE)", node.Name)
+		return
+	case node.Type == sqltypes.Float32:
+		buf.astPrintf(node, "CAST(:%#s AS FLOAT)", node.Name)
+		return
+	case sqltypes.IsDate(node.Type):
+		buf.astPrintf(node, "CAST(:%#s AS DATE)", node.Name)
+		return
+	case node.Type == sqltypes.Time:
+		buf.astPrintf(node, "CAST(:%#s AS TIME)", node.Name)
+		return
+	case node.Type == sqltypes.Timestamp, node.Type == sqltypes.Datetime:
+		buf.astPrintf(node, "CAST(:%#s AS TIMESTAMP)", node.Name)
```

Comment:
I think using a `timestamp :%#s` style literal looks nicer, does that also work? That also automatically has the right precision then. Otherwise we do have to at least also pass in the precision, so like `timestamp(6)` for example if it's a size of `6` of the node. 

Same applies to `time` (and also `date`, but that has no precision concerns).

Reply:
I tried that first, since I started from your code. The issue with that way of printing it is that the string output doesn't parse. `date '2001-01-01'` works fine, but is only valid if the datevalue is a string literal. I tried this on MySQL:

```sql
mysql> set @u = '2001-01-01';
Query OK, 0 rows affected (0.00 sec)

mysql> select date '2001-01-01';
+-------------------+
| date '2001-01-01' |
+-------------------+
| 2001-01-01        |
+-------------------+
1 row in set (0.01 sec)

mysql> select date @u;
ERROR 1064 (42000): You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '@u' at line 1
```

I really want the output of `sqlparser.String()` to be parseable, so I changed it to this.

Reply:
@systay Ah ok, that's a good point. We do need to set the size then for `timestamp` and `time` types to ensure the correct sub second precision.